File: DeflateVB/BitWriter.vb
' public domain

Imports System.IO

Public Class BitWriter
    Public sout As Stream
    Private buf(4095) As Byte
    Private bufp, cur, bit As Integer

    Public Sub WriteByte(b As Byte)
        buf(bufp) = b
        If bufp < 4095 Then
            bufp = bufp + 1
        Else
            sout.Write(buf, 0, buf.Length)
            bufp = 0
        End If
    End Sub

    Public Sub Close()
        If bit > 0 Then
            WriteByte(cur)
            cur = 0
            bit = 0
        End If
        If bufp > 0 Then
            sout.Write(buf, 0, bufp)
            bufp = 0
        End If
        sout.Flush()
    End Sub

    Public Sub WriteBit(b As Boolean)
        If b Then cur = cur Or sl(1, bit, 0)
        If bit < 7 Then
            bit = bit + 1
        Else
            WriteByte(cur)
            cur = 0
            bit = 0
        End If
    End Sub

    Public Sub WriteBits(len%, b%)
        Dim v%, pos%
        If len > 0 Then
            v = cur Or sl(b, bit, 0)
            pos = bit + len
            If pos < 8 Then
                cur = v
                bit = pos
            Else
                WriteByte(v)
                If pos < 16 Then
                    cur = sl(b, bit, 1)
                    bit = pos - 8
                Else
                    WriteByte(sl(b, bit, 1))
                    cur = sl(b, bit, 2)
                    bit = pos - 16
                End If
            End If
        End If
    End Sub

    Public Sub WriteFixedHuffman(b%)
        If b < 144 Then
            WriteBits(8, rev(b + 48))
        ElseIf b < 256 Then
            WriteBit(True)
            WriteBits(8, rev(b))
        ElseIf b < 280 Then
            WriteBits(7, rev((b - 256) << 1))
        ElseIf b < 288 Then
            WriteBits(8, rev(b - 88))
        End If
    End Sub

    Public Sub WriteLen(length%)
        Dim ll%
        ll = litindex(length - 3)
        WriteFixedHuffman(ll)
        WriteBits(litexlens(ll), length - litlens(ll))
    End Sub

    Public Sub WriteDist(d%)
        Dim dl%
        dl = distindex(d - 1)
        WriteBits(5, rev(dl << 3))
        WriteBits(distexlens(dl), d - distlens(dl))
    End Sub
End Class

File: DeflateCS/BitWriter.cs
// public domain

using System;
using System.Collections.Generic;
using System.IO;

namespace DeflateCS
{
    public class BitWriter
    {
        private Stream sout;
        private byte[] buf = new byte[4096];
        private int bufp;
        private byte cur;
        private int bit;

        public BitWriter(Stream sout)
        {
            this.sout = sout;
        }

        public void WriteByte(byte b)
        {
            buf[bufp] = b;
            if (bufp < 4095)
                bufp++;
            else
            {
                sout.Write(buf, 0, buf.Length);
                bufp = 0;
            }
        }

        public void Close()
        {
            if (bit > 0)
            {
                WriteByte(cur);
                cur = 0;
                bit = 0;
            }
            if (bufp > 0)
            {
                sout.Write(buf, 0, bufp);
                bufp = 0;
            }
            sout.Flush();
        }

        public void WriteBit(bool b)
        {
            if (b) cur |= (byte)(1 << bit);
            if (bit < 7)
                bit++;
            else
            {
                WriteByte(cur);
                cur = 0;
                bit = 0;
            }
        }

        public void WriteBits(int len, int b)
        {
            if (len > 0)
            {
                var v = (byte)(cur | Deflate.sl[b, bit, 0]);
                int pos = bit + len;
                if (pos < 8)
                {
                    cur = v;
                    bit = pos;
                }
                else
                {
                    WriteByte(v);
                    if (pos < 16)
                    {
                        cur = Deflate.sl[b, bit, 1];
                        bit = pos - 8;
                    }
                    else
                    {
                        WriteByte(Deflate.sl[b, bit, 1]);
                        cur = Deflate.sl[b, bit, 2];
                        bit = pos - 16;
                    }
                }
            }
        }

        public void WriteFixedHuffman(int b)
        {
            if (b < 144)
                WriteBits(8, Deflate.rev[b + 48]);
            else if (b < 256)
            {
                WriteBit(true);
                WriteBits(8, Deflate.rev[b]);
            }
            else if (b < 280)
                WriteBits(7, Deflate.rev[(b - 256) << 1]);
            else if (b < 288)
                WriteBits(8, Deflate.rev[b - 88]);
        }

        public void WriteLen(int length)
        {
            int ll = Deflate.litindex[length - 3];
            WriteFixedHuffman(ll);
            WriteBits(Deflate.litexlens[ll], length - Deflate.litlens[ll]);
        }

        public void WriteDist(int d)
        {
            int dl = Deflate.distindex[d - 1];
            WriteBits(5, Deflate.rev[dl << 3]);
            WriteBits(Deflate.distexlens[dl], d - Deflate.distlens[dl]);
        }
    }
}
